Fix infinite loop when parsing invalid regular expressions

rozza opened the pull request:
JAVA-2228

jyemin commented:
This looks good, but I'm concerned other bugs are hiding.  I don't see, for example, any tests of the escape sequence path, where there is also no check for EOF.

Also, it's worth putting an EOF check in JsonBuffer to avoid infinite loops, rather than continuing to return -1 over and over.

rozza commented:
>  I don't see, for example, any tests of the escape sequence path, where there is also no check for EOF.

Can you elaborate? I'm not sure what to do here. I checked the other `while(true)` loops and they all handled EOF.  I'll add an EOF check to JsonBuffer as a precaution.

jyemin commented on bson/src/main/org/bson/json/JsonBuffer.java:
```diff
@@ -34,11 +35,19 @@ public void setPosition(final int position) {
-        return (position >= buffer.length()) ? -1 : buffer.charAt(position++);
+        if (eof) {
+            throw new JsonParseException("Trying to read past EOF.");
+    } else if (position >= buffer.length()) {
```
Nit: indentation

rozza replied:
Done.

jyemin replied:
Indentation still looks like it's of

jyemin commented:
Sorry, I mis-read the escape sequence code path.  It's fine.

LGTM

rozza commented:
rebased and merged.

rozza commented:
Had to revert as I failed to check the other test cases.  

Updated a test to not try and read past eof.
When unreading we now always set eof to false
Updated the JsonReader to not try to read past eof.

rozza commented:
Also do you want me to backport to 2.14.x ?

jyemin commented on bson/src/main/org/bson/json/JsonReader.java:
```diff
@@ -226,6 +226,8 @@ public BsonType readBsonType() {
         }
         if (noValueFound) {
             throw new JsonParseException("JSON reader was expecting a value but found '%s'.", token.getValue());
+        } else if (token.getType() == JsonTokenType.END_OF_FILE) {
```
This seems to make a bad situation worse.  I see that this allows JsonReaderTest#testEndOfFile1 to pass, but is there any good reason why it _should_ pass?  The test subject is

    "{ test : "

so why should JsonReader generate BsonType.END_OF_DOCUMENT when there is no closing curly brace?

rozza replied:
Fixed

jyemin commented:
lgtm

rozza commented:
rebased and merged.
